File: package/android/src/main/java/com/margelo/nitro/nitromaps/MarkerDescriptor.kt
package com.margelo.nitro.nitromaps

/**
 * Marker data as the native store holds it.
 *
 * Nitrogen used to generate these classes from the `markers` view prop. Markers
 * now reach native code as packed batches (see [MarkerBatchDecoder]), so no
 * spec references them and they live here instead. Field names and types match
 * the TypeScript `MarkerDescriptor`. Should a Nitro spec reference
 * `MarkerDescriptor` again, nitrogen would generate a conflicting class and
 * this file has to go.
 */
data class MarkerImage(
  val uri: String,
  val width: Double?,
  val height: Double?,
  val scale: Double?,
)

/** Anchor point on the marker image (0..1). */
data class MarkerAnchor(
  val x: Double,
  val y: Double,
)

/** Point offset in density-independent pixels. */
data class MarkerPoint(
  val x: Double,
  val y: Double,
)

data class MarkerDescriptor(
  val id: String,
  val coordinate: Coordinate,
  val title: String?,
  val subtitle: String?,
  val draggable: Boolean?,
  val clusterable: Boolean?,
  val image: MarkerImage?,
  val anchor: MarkerAnchor?,
  val centerOffset: MarkerPoint?,
  val rotation: Double?,
  val flat: Boolean?,
  val opacity: Double?,
  val enteringAnimation: OverlayEnteringAnimationDescriptor?,
)

File: package/ios/MarkerDescriptor.swift
/// Marker data as the native store holds it.
///
/// Nitrogen used to generate these structs from the `markers` view prop. Markers
/// now reach native code as packed batches (see `MarkerBatchDecoder`), so no
/// spec references them and they live here instead. Field names and types
/// match the TypeScript `MarkerDescriptor`. Should a Nitro spec reference
/// `MarkerDescriptor` again, nitrogen would generate a conflicting type and
/// this file has to go.
struct MarkerImage {
  var uri: String
  var width: Double?
  var height: Double?
  var scale: Double?
}

/// Anchor point on the marker image (0..1).
struct MarkerAnchor {
  var x: Double
  var y: Double
}

/// Point offset in density-independent pixels.
struct MarkerPoint {
  var x: Double
  var y: Double
}

struct MarkerDescriptor {
  var id: String
  var coordinate: Coordinate
  var title: String?
  var subtitle: String?
  var draggable: Bool?
  var clusterable: Bool?
  var image: MarkerImage?
  var anchor: MarkerAnchor?
  var centerOffset: MarkerPoint?
  var rotation: Double?
  var flat: Bool?
  var opacity: Double?
  var enteringAnimation: OverlayEnteringAnimationDescriptor?
}
